File: src/main/scala/uk/co/ridentbyte/model/Command.scala
package uk.co.ridentbyte.model

import java.util.Random

import scala.util.matching.Regex
import scala.collection.JavaConverters._

/**
  * Handles commands in the format: #{(.*)}
  *
  * @param command - Raw command
  */
case class Command(private val command: String) {

  /**
    * Extract contents from command
    */
  def contents: String = {
    "#\\{(.*)\\}".r
      .findFirstMatchIn(Option(command).getOrElse(""))
      .map(_.group(1))
      .getOrElse(command)
  }

  def rawCommand: String = command

  def processWith(constants: Map[String, () => String], vocabulary: Vocabulary, prev: Option[String] = None, r: Random = new Random()): String = {
    val rawContents = prev.getOrElse(contents)
    val optDirectlyMatchedVar = constants.find(_._1 == rawContents)
    if (optDirectlyMatchedVar.isDefined) {
      optDirectlyMatchedVar.get._2()
    } else {
      // Not a variable, process functions

      val functions = List[Regex](
        "(?:lower\\((.*)\\))".r,
        "(?:upper\\((.*)\\))".r,
        "(?:capitalise\\((.*)\\))".r,
        "(?:random\\((.*)\\))".r,
        "(?:lorem\\((.*)\\))".r,
        "(?:randomBetween\\((.*)\\))".r
      )

      val lowerFunction = "(?:lower\\((.*)\\))".r
      val upperFunction = "(?:upper\\((.*)\\))".r
      val capitaliseFunction = "(?:capitalise\\((.*)\\))".r
      val randomFunction = "(?:random\\((.*)\\))".r
      val loremFunction = "(?:lorem\\((.*)\\))".r
      val randomBetween = "(?:randomBetween\\((.*)\\))".r

      val rawString = "\"[^\"]*\""

      rawContents match {
        case lowerFunction(v) if v.matches(rawString) => v.replace("\"", "").toLowerCase()
        case lowerFunction(v) if !v.matches(rawString) => processWith(constants, vocabulary, Some(v)).toLowerCase
        case upperFunction(v) if v.matches(rawString) => v.replace("\"", "").toUpperCase()
        case upperFunction(v) if !v.matches(rawString) => processWith(constants, vocabulary, Some(v)).toUpperCase
        case capitaliseFunction(v) if v.matches(rawString) => v.replace("\"", "").capitalize
        case capitaliseFunction(v) if !v.matches(rawString) => processWith(constants, vocabulary, Some(v)).capitalize
        case loremFunction(n) =>
          val argumentType = getArgumentTypes(List(n), constants, functions).head
          if (argumentType.isInstanceOf[IntLiteral]) {
            vocabulary.loremIpsum.first(n.toInt).asScala.mkString(" ")
          } else if (argumentType.isInstanceOf[Function]) {
            val processedValue = processWith(constants, vocabulary, Some(n))
            val processedValueArgumentType = getArgumentTypes(List(processedValue), constants, functions)
            if (processedValueArgumentType.head.isInstanceOf[IntLiteral]) {
              vocabulary.loremIpsum.first(processedValue.toInt).asScala.mkString(" ")
            } else {
              rawContents
            }
          } else {
            rawContents
          }

        case randomFunction(v) =>
          // TODO - need to improve text splitting here
          // For example we want #{random(random(1, 2), random(3, 4))} to be valid
          val splitValues = v.split(",").map(_.trim).toList
          val argumentTypes = getArgumentTypes(splitValues, constants, functions)
          val containsInvalidTypes = argumentTypes.count(n => n.isInstanceOf[InvalidLiteral]) > 0
          if (!containsInvalidTypes) {
            val processedValues = splitValues.map { v => processWith(constants, vocabulary, Some(v)) }
            processedValues(r.nextInt(processedValues.length))
          } else {
            rawContents
          }

        case randomBetween(v) =>
          val splitValues = v.split(",").map(_.trim).toList
          val argumentTypes = getArgumentTypes(splitValues, constants, functions)
          val containsInvalidTypes = argumentTypes.count(n => n.isInstanceOf[InvalidLiteral]) > 0
          if (!containsInvalidTypes && splitValues.length == 2) {
            try {
              val processedValues = splitValues.map { v => processWith(constants, vocabulary, Some(v)) }
              val allPossibleValues = processedValues.head.toInt.to(processedValues.last.toInt)
              allPossibleValues(r.nextInt(allPossibleValues.length)).toString
            } catch {
              case _: Exception => rawContents
            }
          } else {
            rawContents
          }

        case v if v.matches(rawString) => v.replace("\"", "")
        case v => v
      }
    }
  }

  def getArgumentTypes(args: List[String], constants: Map[String, () => String], functions: List[Regex]): List[ArgumentType] = {
    args.map {
      case a if a.matches("\"[^\"]*\"") => StringLiteral(a)
      case i if i.matches("[\\-]?[0-9]+") => IntLiteral(i)
      case c if constants.contains(c) => Constant(c)
      case f if functions.exists(f2 => f.matches(f2.regex)) => Function(f)
      case n => InvalidLiteral(n)
    }
  }

  def isRawInt(s: String): Boolean = s.matches("[0-9]+")

}

class ArgumentType(value: String)
case class Function(value: String) extends ArgumentType(value)
case class Constant(value: String) extends ArgumentType(value)
case class StringLiteral(value: String) extends ArgumentType(value)
case class IntLiteral(value: String) extends ArgumentType(value)
case class InvalidLiteral(value: String) extends ArgumentType(value)

File: src/main/java/uk/co/ridentbyte/model/Command.java
package uk.co.ridentbyte.model;

import java.util.Arrays;
import java.util.List;
import java.util.Map;
import java.util.Random;
import java.util.concurrent.ThreadLocalRandom;
import java.util.function.Supplier;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import java.util.stream.Collectors;


// TODO
// - This class doesn't handle the "#{ }" container anymore so this needs to be
//   moved up to the calling class
// - This class needs to detect the different types the different functions take
//   so it can act differently based on argument types.
//   class ArgumentType(value: String), case class Function(value: String), case class Constant(value: String)
public class Command {

    private String command;

    private Pattern lowerFunction = Pattern.compile("(?:lower\\((.*)\\))");
    private Pattern upperFunction = Pattern.compile("(?:upper\\((.*)\\))");
    private Pattern capitaliseFunction = Pattern.compile("(?:capitalise\\((.*)\\))");
    private Pattern randomFunction = Pattern.compile("(?:random\\((.*)\\))");
    private Pattern loremFunction = Pattern.compile("(?:lorem\\((.*)\\))");
    private Pattern randomBetweenFunction = Pattern.compile("(?:randomBetween\\((.*)\\))");
    private Pattern rawString = Pattern.compile("(?:^\"(.*)\"$)");

    public Command(String command) {
        this.command = command;
    }

    public String process(Map<String, Supplier<String>> variables, Vocabulary vocabulary) {
        return process(variables, vocabulary, command);
    }

    private String process(Map<String, Supplier<String>> variables, Vocabulary vocabulary, String stackedCommand) {
        if (variables.containsKey(stackedCommand)) {
            String value = variables.get(stackedCommand).get();
            return process(variables, vocabulary, value);
        } else if (stackedCommand.matches(lowerFunction.pattern())) {
            Matcher matcher = lowerFunction.matcher(stackedCommand);
            matcher.find();
            return process(variables, vocabulary, matcher.group(1)).toLowerCase();
        } else if (stackedCommand.matches(upperFunction.pattern())) {
            Matcher matcher = upperFunction.matcher(stackedCommand);
            matcher.find();
            return process(variables, vocabulary, matcher.group(1)).toUpperCase();
        } else if (stackedCommand.matches(capitaliseFunction.pattern())) {
            Matcher matcher = capitaliseFunction.matcher(stackedCommand);
            matcher.find();
            return capitalise(process(variables, vocabulary, matcher.group(1)));
        } else if (stackedCommand.matches(rawString.pattern())) {
            Matcher matcher = rawString.matcher(stackedCommand);
            matcher.find();
            return process(variables, vocabulary, matcher.group(1));
        } else if (stackedCommand.matches(randomFunction.pattern())) {
            Matcher matcher = randomFunction.matcher(stackedCommand);
            matcher.find();
            List<String> processedValues =
                    Arrays.stream(matcher.group(1).split(","))
                            .map(v -> process(variables, vocabulary, v.trim()))
                            .collect(Collectors.toList());
            return random(processedValues);
        } else if (stackedCommand.matches(loremFunction.pattern())) {
            Matcher matcher = loremFunction.matcher(stackedCommand);
            matcher.find();
            String processedArguments = process(variables, vocabulary, matcher.group(1));
            if (processedArguments.matches("[0-9]+")) {
                return lorem(Integer.parseInt(processedArguments), vocabulary.loremIpsum);
            } else {
                return command;
            }
        } else if (stackedCommand.matches(randomBetweenFunction.pattern())) {
            Matcher matcher = randomBetweenFunction.matcher(stackedCommand);
            matcher.find();
            String[] splitArgs = matcher.group(1).split(",");
            if (splitArgs.length == 2) {
                List<String> processedArgs = Arrays.stream(splitArgs)
                        .map((arg) -> process(variables, vocabulary, arg.trim()))
                        .collect(Collectors.toList());
                if (processedArgs.get(0).matches("[0-9]+") && processedArgs.get(1).matches("[0-9]+")) {
                    int min = Math.min(Integer.parseInt(processedArgs.get(0)), Integer.parseInt(processedArgs.get(1)));
                    int max = Math.max(Integer.parseInt(processedArgs.get(0)), Integer.parseInt(processedArgs.get(1)));
                    return Integer.toString(ThreadLocalRandom.current().nextInt(min, max + 1));
                } else {
                    return command;
                }
            } else {
                return command;
            }
        } else {
            return stackedCommand;
        }
    }

    private String capitalise(String value) {
        return value.substring(0, 1).toUpperCase() + value.substring(1);
    }

    private String random(List<String> values) {
        return values.get(new Random().nextInt(values.size()));
    }

    private String lorem(int i, Words loremIpsum) {
        return String.join(" ", loremIpsum.first(i));
    }

    public String getCommand() {
        return command;
    }
}
